### PR TITLE
[datareposrc] Add function to read flexible tensors

### PR DESCRIPTION
**Pull request by songgot**
- If the tensor format of the caps of the JSON file is flexible,
      Reads a sample using the flexible meta information (sample_offset, tensor_size, tensor_count)
      of the JSON file and appends memory to GstBuffer as many as the number of flexible tensors.
    - Add unit test
    - Add checking if it is a flexible tensor after writing data to GstMemory
    
    - Reference
      * The start offset for reading is sample_offset(sample size).
      * Save each flexible tensor stored in a sample to a gstbuffer according to each
        tensor_size, tensor_size can get tensor_size field in JSON file.
      * A shuffled index is mapped to an index of sample_offset field in JSON file.
      * A shuffled index is also mapped to an index of tensor_count field in JSON file
      * The index value is a number of cumulative tensors, so it is mapped to an index
        of tensor_size field in JSON file

**Self evaluation:**
1. Build test: [*]Passed [ ]Failed [ ]Skipped
2. Run test: [*]Passed [ ]Failed [ ]Skipped

**Reference**
#4122 
If a shuffle index is 2, Let's see (1),(2),(3) how it works.
$cat flexible.json
```
{
  "gst_caps" : "other/tensors, format=(string)flexible, framerate=(fraction)10/1",
  "total_samples" : 9,
  "sample_offset" : [
    0, 
    152320, 
    1304576, ---------> (1) shuffle index 2, so do lseek(fd, 1304576, SET_SET)
    2456832,
    2609152,
    3761408,
    4913664,
    5065984,
    5527040
  ],
  "tensor_size" : [
    76160, 
    76160, 
    921728, 
    230528, 
    230528, --------> (3)tensor_count  is "4" and num_tensor is 2 when shuffle index 2, so datareposrc uses
    921728,          gst_buffer_append_memory() to append 230528byte and 921728byte memory to a Gstbuffer.
    76160,
    76160,
    921728,
    230528,
    230528,
    921728,
    76160,
    76160,
    230528,
    230528,
    921728,
    921728 
  ] ----------------> total count of tensor_size array is "18"
  "tensor_count" : [
    0,
    2,
    4, -------------> (2) shuffle index 2, so num_tensors is 2 (6 - 4), and current tensor_count  is "4"
    6,                        , "4" is index of tensor_size
    8,
    10,
    12,
    14,
    16 
  ] 
}
```

### CODE REVIEW

File: gst/datarepo/gstdatareposrc.c [hunk 1]
```diff
@@ -526,17 +541,176 @@ gst_data_repo_src_read_tensors (GstDataRepoSrc * src, GstBuffer ** buffer)
...
+    gst_buffer_unref (buf);
+    return GST_FLOW_ERROR;
+  }
+eos:
+  {
+    GST_DEBUG ("EOS");
+    gst_memory_unmap (mem, &info);
+    gst_buffer_unref (buf);
+    return GST_FLOW_EOS;
+  }
```
Comment (gichan-jang) [inline gst/datarepo/gstdatareposrc.c:601]: Let's remove duplicated code

Comment (songgot) [reply gst/datarepo/gstdatareposrc.c:601]: okay, thank you.

Comment (songgot) [reply gst/datarepo/gstdatareposrc.c:601]: It Cannot merge because the return values are different.

File: gst/datarepo/gstdatareposrc.c [hunk 2]
```diff
@@ -526,66 +543,218 @@ gst_data_repo_src_read_tensors (GstDataRepoSrc * src, GstBuffer ** buffer)
...
 
-      src->read_position += ret;
-      src->offset += ret;
+      src->read_position += read_size;
+      src->fd_offset += read_size;
     }
 
-    if (mem[i])
-      gst_memory_unmap (mem[i], &info[i]);
+    if (mem)
```
Comment (jaeyun-jung) [inline gst/datarepo/gstdatareposrc.c:584]: IMO this cond is unnecessary. gst_memory_map() at line 520 checks null case (allocation failure)

Comment (songgot) [reply gst/datarepo/gstdatareposrc.c:584]: Thank you for your review. :)

File: gst/datarepo/gstdatareposrc.c [hunk 3]
```diff
@@ -526,66 +543,218 @@ gst_data_repo_src_read_tensors (GstDataRepoSrc * src, GstBuffer ** buffer)
...
   return GST_FLOW_OK;
 
-could_not_read:
-  {
-    GST_ELEMENT_ERROR (src, RESOURCE, READ, (NULL), GST_ERROR_SYSTEM);
-    gst_memory_unmap (mem[0], &info[0]);
-    gst_buffer_unref (buf);
-    return GST_FLOW_ERROR;
+error:
+  gst_memory_unmap (mem, &info);
```
Comment (jaeyun-jung) [inline gst/datarepo/gstdatareposrc.c:595]: suggestion: leak case if failed to map memory at line 518 mem = gst_allocator_alloc();

```
GstMemory *mem = NULL;
...

error:
  if (mem) {
    gst_memory_unmap (mem, &info);
    gst_memory_unref (mem);
  }

  gst_buffer_unref (buf);
  return ret;
```

Comment (songgot) [reply gst/datarepo/gstdatareposrc.c:595]: ```mem``` does not need to ```gst_memory_unref (mem)```.

Comment (songgot) [reply gst/datarepo/gstdatareposrc.c:593]: Okay, I got it!!, thank you.

File: tests/nnstreamer_datarepo/unittest_datareposrc.cc [hunk 4]
```diff
@@ -45,8 +45,10 @@ bus_callback (GstBus *bus, GstMessage *message, gpointer data)
 {
   switch (GST_MESSAGE_TYPE (message)) {
     case GST_MESSAGE_EOS:
+    case GST_MESSAGE_ERROR:
       g_main_loop_quit ((GMainLoop *) data);
       break;
+      break;
```
Comment (jaeyun-jung) [inline tests/nnstreamer_datarepo/unittest_datareposrc.cc:51]: unnecessary?

Comment (songgot) [reply tests/nnstreamer_datarepo/unittest_datareposrc.cc:51]: When I do git rebase, the merge seems to be wrong.
Thank you.

**Review by gichan-jang — APPROVED** [review]
LGTM

**Comment by songgot** [comment]
Thank you for your review. special thank @jaeyun-jung, @gichan-jang

**Pull request merged**